src/hidra/receiver/asapo_transfer.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import logging
+import signal
+from threading import Event
+from time import sleep
+
+from hidra import Transfer, __version__, generate_filepath
+from plugins.asapo_producer import AsapoWorker
+
+logger = logging.getLogger(__name__)
+
+
+class Stopped(Exception):
+    """Raised when a run is stopped."""
+    pass
+
+
+class AsapoTransfer:
+    def __init__(self, asapo_worker, signal_host, target_host, target_port, target_dir, reconnect_timout):
+
+        self.signal_host = signal_host
+        self.targets = [[target_host, target_port, 1]]
+        self.target_dir = target_dir
+        self.asapo_worker = asapo_worker
+        self.reconnect_timout = reconnect_timout
+        self.query = Transfer("STREAM_METADATA", self.signal_host)
+        self.stop_run = Event()
+
+    def run(self):
+        if self.stop_run.is_set():
+            raise Stopped
+        try:
+            self.query.initiate(self.targets)
+            self.query.start()
+            self._run()
+        finally:
+            self.query.stop()
+
+    def _run(self):
+        while not self.stop_run.is_set():
+            [metadata, data] = self.query.get(timeout=self.reconnect_timout*1000)
+            if metadata is not None:
+                try:
+                    local_path = generate_filepath(self.target_dir, metadata)
+                    logger.info("Send file {local_path}".format(local_path=local_path))
+                    self.asapo_worker.send_message(local_path, metadata)
+                except Exception as e:
+                    logger.error("Transmission does not succeed. {e}. File ignored".format(e=str(e)))
+
+    def stop(self):
+        logger.info("Runner is stopped.")
+        self.stop_run.set()
+
+
+def main():
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+                                     description='Transfer files metadata to ASAPO')
+    parser.add_argument('--endpoint', type=str, help='ASAPO produces endpoint')
+    parser.add_argument('--beamtime', type=str, help='ASAPO produces beamtime')
+    parser.add_argument('--beamline', type=str, help='ASAPO produces beamline')
+    parser.add_argument('--default-data-source', type=str, help='ASAPO data_source')
+    parser.add_argument('--token', type=str, help='ASAPO produces token')
+    parser.add_argument('--n_threads', type=int, help='Number of threds for ASAPO producer',
+                        default=1)
+    parser.add_argument('--start_file_idx', type=int, help='Starting file index',
+                        default=1)
+    parser.add_argument('--file_regex', type=str, help='Template to file path, which includes `stream` and `file_idx`',
+                        default=".*raw/(?P<scan_id>.*)/\w*_(?P<file_idx_in_scan>.*).h5")
```

Comment:
I think this is not sufficient. There might be files from multiple scans with different prefixes in the same directory, i.e., the files `raw/foo1/bar1_1.h5`, `raw/foo1/bar2_1.h5`, `raw/foo2/bar1_1.h5`, and `raw/foo2/bar2_1.h5` should end up in different streams. This would be easy if Asapo supports stream names containing `/` (which this regex assumes as well for files in nested directories like `raw/foo/bar/baz_1.h5`).

We should also discuss what to do with the master file.

Reply:
Add more test on different files. File regex is changed.